### ISSUE TITLE
Total COB overcount when Carbs in Health change

### ISSUE THREAD

**Issue by CrushingT1D** [LoopKit/Loop#243]
If carbs are entered into HealtKit and then modified a short time later, Loop will count both the first entry and revised entry in total COB. This can present a dangerous situation indicating more carbs ate than in reality.

This is incredibly easy to do with a 3rd party app like myfitnesspal that writes carbs to HealtKit. Example to reproduce is to add a breakfast item with carbs in myfitnesspal, then get distracted for 2 minutes, then add a second item with carbs. The entry in HealthKit will be updated with the full meal that happened at the same time, however Loop with consider total COB to be the first item and then the full breakfast (first and second item). The details in CarbKit show up correctly, however the total COB is incorrect per the screenshots below.

![img_6725](https://cloud.githubusercontent.com/assets/18058098/19849547/f57ae228-9fa7-11e6-9014-735a3f702b36.PNG)
![img_6723](https://cloud.githubusercontent.com/assets/18058098/19849549/f8bb7650-9fa7-11e6-9e30-f312123a2016.PNG)
![img_6722](https://cloud.githubusercontent.com/assets/18058098/19849550/fce18184-9fa7-11e6-955e-fa1e3a43b7ea.PNG)

The only way I have figured out how to fix this issues is to kill Loop app right away and reopen it before an artificial high temp gets out of hand.

**Comment by CrushingT1D** [comment]
I have been able to repeat this several times on queue.
I have verified that after this happens that getRecentCarbEntries does not display both sets of carbs (as printed entires here https://github.com/LoopKit/LoopKit/blob/e8035f39c39c5a82ce7fb1bf93e13b7d5e3bad72/CarbKit/CarbStore.swift#L699)

**Comment by CrushingT1D** [comment]
It might be worth ceasing 3rd party carb entry access in Loop until this can be fixed.

**Comment by ps2** [comment]
Yes, you're right this is important. I'll either get a fix out, or disable carb entry.

**Comment by ps2** [comment]
I'm having trouble reproducing this.  I think I actually did reproduce this morning, but then ran out of time to work on it.  Tonight I've been trying a variety of adds/deletes and Loop seems to catch them all. Any tips for reproducing?

**Comment by CrushingT1D** [comment]
Try testing only a meal or snack that is close to the current time of day. Otherwise they may not get added by myfitnesspal to healthkit. Demo video attached showing a case of not causing the bug and then reproducing it @ps2 

[IMG_1317.MOV.mov.zip](https://github.com/LoopKit/Loop/files/633407/IMG_1317.MOV.mov.zip)

**Comment by ps2** [comment]
@CrushingT1D Are you still able to reproduce this? I just tried for another hour using the LoopKit example app, and Loop, and MyFitnessPal.  Did they maybe update the app, or I'm wondering if something in Loop changed. I did see one time where the display was off for about a second, and then it corrected.

If you are able to reproduce it, can you wait five minutes, and see if it actually makes a temp basal decision using the data?  From what I see digging into the code and what you show on your video, it might just be a display issue.

**Comment by ps2** [comment]
For anyone who can reproduce this issue, I've added some changes to LoopKit that may fix a race condition.  If you could update your cartfile to 

```

**Comment by CrushingT1D** [comment]
@ps2 I have tested the carbstore-async-cleanup for 4 days using myfitnesspal and I have not been able to duplicate the issue that I saw in the past. It is my belief that the code changes here https://github.com/LoopKit/LoopKit/commit/123a4d2ba98aba8adb102c78a3c62e670ea330e7
have fixed the double counting issue as part of that race condition fix. cc/ @shanselman 

I also think we should add a section to the guide that talks about 3rd part app logging. There are some things to be aware of with timestamps those apps use. cc/ @Kdisimone

**Comment by shanselman** [comment]
:shipit:

**Comment by CrushingT1D** [comment]
Throwing the breaks on this one. I just came across the bug in this branch.
Case where I duplicated it
- Entered breakfast (3 items) into myfitnesspal, when to HealthKit and then Loop without issue. Showed 8 carbs. 
- Added a 4th item which had 1 carb. Loop updated fine with 9 total.
- Changed one of the 4 items to a smaller quantity. Should have dropped from 9 to 5 carbs for the meal. 
- Loop showed 14 COB. Clicked through and only 5 carbs shows in the list of carbs, but total on board still shows 14.
- force closed Loop, reopened and shows 5 COB as it should

I will try to recreate this (later today/tonight) using the debug code of the LoopKit sample app and post the output.
